onnxruntime/core/platform/windows/env.cc

```diff
@@ -212,52 +221,121 @@ class WindowsEnv : public Env {
-        ORT_THROW("Fatal error: 0 count processors from GetSystemInfo");
+        return std::max(1, static_cast<int>(std::thread::hardware_concurrency()/2));
       }
       // This is the number of logical processors in the current group
-      return sysInfo.dwNumberOfProcessors;
+      return std::max(1, static_cast<int>(sysInfo.dwNumberOfProcessors/2));
     }
-    int processorCoreCount = 0;
-    int count = (int)(returnLength / sizeof(SYSTEM_LOGICAL_PROCESSOR_INFORMATION));
-    for (int i = 0; i != count; ++i) {
+
+    SYSTEM_LOGICAL_PROCESSOR_INFORMATION* buffer =
+        reinterpret_cast<SYSTEM_LOGICAL_PROCESSOR_INFORMATION*>(logical_processor_info->buffer_.get());
+    const size_t count = logical_processor_info->count_;
+    int phys_cores = 0;
+    for (size_t i = 0; i < count; ++i) {
       if (buffer[i].Relationship == RelationProcessorCore) {
-        ++processorCoreCount;
+        phys_cores++;
       }
     }
-    if (!processorCoreCount)
-      ORT_THROW("Fatal error: 0 count processors from GetLogicalProcessorInformation");
-    return processorCoreCount;
-  }
-
-  std::vector<size_t> GetThreadAffinityMasks() const override {
-    auto generate_vector_of_n = [](int n) {
-      std::vector<size_t> ret(n);
-      std::iota(ret.begin(), ret.end(), 0);
+    return std::max(1, phys_cores);
+  }
+
+  std::vector<ThreadOptions::ThreadAffinity> GetThreadAffinityMasks() const override {
+    // Default mask generation if everything else fails.
+    auto generate_vector_of_n = [](unsigned int num_logical_proc) {
+      // The return values will be converted into a 64-bit
+      // mask at the site of the SetThreadAffinityMask() call, so we need to make sure
+      // that num_logical_proc does not exceed 64, which is the max number of processors
+      // supported by SetThreadAffinityMask(). We assume two logical processors per
+      // physical core.
+      num_logical_proc = std::clamp(num_logical_proc, 1U, 64U);
+      std::vector<ThreadOptions::ThreadAffinity> ret;
+      if (num_logical_proc == 1U) {
+        ThreadOptions::ThreadAffinity th_aff{{0}};
+        ret.push_back(std::move(th_aff));
+        return ret;
+      }
+      ret.reserve(num_logical_proc / 2);
+      for (unsigned int c = 0; c < num_logical_proc; c += 2) {
+        ThreadOptions::ThreadAffinity th_aff{{static_cast<int>(c), static_cast<int>(c + 1)}};
+        ret.push_back(std::move(th_aff));
+      }
       return ret;
     };
-    // Indeed 64 should be enough. However, it's harmless to have a little more.
-    SYSTEM_LOGICAL_PROCESSOR_INFORMATION buffer[256];
-    DWORD returnLength = sizeof(buffer);
-    if (GetLogicalProcessorInformation(buffer, &returnLength) == FALSE) {
+
+    std::vector<ThreadOptions::ThreadAffinity> ret;
+
+    auto logical_processor_info = FetchLogicalProcessorInfo();
+    if (!logical_processor_info.has_value()) {
       return generate_vector_of_n(std::thread::hardware_concurrency());
```

Comment:
>return generate_vector_of_n(std::thread::hardware_concurrency()); [](http://example.com/codeflow?start=6&length=65)

The error handling of both windows and posix is questionable - if for some reason, either cpu_info/windows API calls failed to get us some valid result, we should prefer to not set mask at all instead generating masks "by default", which are like to also fail during affinity-setting phase.

Reply:
I agree with you.